src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,207 @@
+  public Command shooterIdle() {
+    return run(() -> {
+          setShooterSpeeds(Constants.ShooterConstants.SHOOTER_IDLE_SPEEDS);
+        })
+        .withName("Idle shooter command");
+  }
+
+  // Checks if shooter is ready.
+  public void isShooterReady(Pair<Double, Double> targetSpeeds) {
+    double targetSpeedTop = targetSpeeds.getFirst();
+    double targetSpeedBottom = targetSpeeds.getSecond();
+
+    topMotorTargetSpeedEntry.log(targetSpeedTop);
+    bottomMotorTargetSpeedEntry.log(targetSpeedBottom);
+
+    double minSpeedToleranceTop =
+        targetSpeedTop * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedToleranceTop =
+        targetSpeedTop * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double minSpeedToleranceBottom =
+        targetSpeedBottom * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedToleranceBottom =
+        targetSpeedBottom * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+
+    double speedTop = shooterTopMotor.getEncoder().getVelocity();
+    double speedBottom = shooterBottomMotor.getEncoder().getVelocity();
+
+    topMotorSpeedEntry.log(speedTop);
+    bottomMotorSpeedEntry.log(speedBottom);
+
+    if ((minSpeedToleranceTop < speedTop && speedTop < maxSpeedToleranceTop)
+        && (minSpeedToleranceBottom < speedBottom && speedBottom < maxSpeedToleranceBottom)
+            == true) {
+      // Request note from feeder.
+    }
+  }
+
+  // Speed in RPM. Top is index 0, bottom is index 1.
+  public Pair<Double, Double> setShooterSpeeds(Pair<Double, Double> speeds) {
+    Pair<Double, Double> targetSpeeds = speeds;
+
+    shooterTopMotor
+        .getPIDController()
+        .setReference(speeds.getFirst(), CANSparkBase.ControlType.kVelocity);
+    shooterBottomMotor
+        .getPIDController()
+        .setReference(speeds.getSecond(), CANSparkBase.ControlType.kVelocity);
+
+    return targetSpeeds;
+  }
+
+  // Top is index 0, bottom is index 1.
+  public static Pair<Double, Double> calculateShooterSpeeds(double distance) {
+    Pair<Double, Double> speeds;
+    Double topSpeed = 0d;
+    Double bottomSpeed = 0d;
+    double distanceProportion;
+
+    // If distance below minimum, set speed to minimum.
+    if (distance < speakerConfigList[0].getDistanceInInches()) {
+      topSpeed = speakerConfigList[0].getSpeedTopInRPM();
+      bottomSpeed = speakerConfigList[0].getSpeedBottomInRPM();
+      speeds = new Pair(topSpeed, bottomSpeed);
+      return speeds;
+    } else if (distance >= speakerConfigList[0].getDistanceInInches()
+        && distance <= speakerConfigList[speakerConfigList.length - 1].getDistanceInInches()) {
+      // A linear search which determines which points the input distance falls between. May be
+      // converted to a binary search if there are many points
+      for (int i = 0; i < speakerConfigList.length - 1; i++) {
+        // If distance above maximum, set speed to maximum.
```

Comment:
I believe this comment applies to the block of code after this if block, currently starting on line 152.

Reply:
Resolved in 47b8ded.